File: Motocitizen/src/motocitizen/ui/fragments/SettingsFragment.java
package motocitizen.ui.fragments;

import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.support.annotation.NonNull;

import motocitizen.main.R;
import motocitizen.router.Router;
import motocitizen.user.User;
import motocitizen.utils.LocationUtils;
import motocitizen.utils.Preferences;
import motocitizen.utils.ToastUtils;

public class SettingsFragment extends PreferenceFragment {
    private static final int PREFERENCES = R.xml.preferences;

    private Preference notificationDistPreference;
    private Preference notificationAlarmPreference;
    private final Preferences preferences = Preferences.INSTANCE;

    private Preference buttonAuth;
    private Preference buttonSound;
    private Preference showAcc;
    private Preference showBreak;
    private Preference showSteal;
    private Preference showOther;
    private Preference hoursAgo;
    private Preference maxNotifications;
    private Preference useVibration;
    private Preference notificationSoundPreference;
    private Preference authPreference;

    String login = preferences.getLogin();

    @Override
    public void onResume() {
        super.onResume();
        setPreferenceScreen(null);
        addPreferencesFromResource(PREFERENCES);
        bindPreferences();
        setUpListeners();
        update();
    }

    private void update() {
        authPreference.setSummary(login.length() > 0 ? User.INSTANCE.getRoleName() + ": " + login : User.INSTANCE.getRoleName());
        maxNotifications.setSummary(String.valueOf(preferences.getMaxNotifications()));
        hoursAgo.setSummary(String.valueOf(preferences.getHoursAgo()));
        notificationSoundPreference.setSummary(preferences.getSoundTitle());
        notificationDistPreference.setSummary(String.valueOf(preferences.getVisibleDistance()));
        notificationAlarmPreference.setSummary(String.valueOf(preferences.getAlarmDistance()));
    }

    private void setUpListeners() {
        notificationDistPreference.setOnPreferenceChangeListener(this::distanceListener);
        notificationAlarmPreference.setOnPreferenceChangeListener(this::distanceListener);
        maxNotifications.setOnPreferenceChangeListener(this::maxNotificationsListener);
        hoursAgo.setOnPreferenceChangeListener(this::hoursAgoListener);
        useVibration.setOnPreferenceChangeListener(this::vibrationListener);
        buttonSound.setOnPreferenceClickListener(this::soundButtonPressed);
        buttonAuth.setOnPreferenceClickListener(this::authButtonPressed);
        showAcc.setOnPreferenceChangeListener(this::visibleListener);
        showBreak.setOnPreferenceChangeListener(this::visibleListener);
        showSteal.setOnPreferenceChangeListener(this::visibleListener);
        showOther.setOnPreferenceChangeListener(this::visibleListener);
    }

    private boolean maxNotificationsListener(Preference preference, Object newValue) {
        preference.setSummary((String) newValue);
        return true;
    }

    private boolean hoursAgoListener(Preference preference, Object newValue) {
        if (newValue.equals("0")) newValue = "1";
        preference.setSummary(newValue.toString());
        return true;
    }

    private boolean vibrationListener(Preference preference, Object newValue) {
        preferences.putBoolean(preference.getKey(), (boolean) newValue);
        return true;
    }

    private boolean authButtonPressed(Preference stub) {
        Router.INSTANCE.goTo(getActivity(), Router.Target.AUTH);
        return true;
    }

    private boolean soundButtonPressed(Preference stub) {
        getFragmentManager().beginTransaction().replace(android.R.id.content, new SelectSoundFragment()).commit();
        return true;
    }

    private void bindPreferences() {
        notificationDistPreference = findPreference(preferences.getPreferenceName("distanceShow"));
        notificationAlarmPreference = findPreference(preferences.getPreferenceName("distanceAlarm"));
        buttonAuth = findPreference(getResources().getString(R.string.settings_auth_button));
        buttonSound = findPreference(getResources().getString(R.string.notification_sound));
        showAcc = findPreference(preferences.getPreferenceName("showAcc"));
        showBreak = findPreference(preferences.getPreferenceName("showBreak"));
        showSteal = findPreference(preferences.getPreferenceName("showSteal"));
        showOther = findPreference(preferences.getPreferenceName("showOther"));
        hoursAgo = findPreference(preferences.getPreferenceName("hoursAgo"));
        maxNotifications = findPreference(preferences.getPreferenceName("maxNotifications"));
        useVibration = findPreference(preferences.getPreferenceName("useVibration"));
        notificationSoundPreference = findPreference(getResources().getString(R.string.notification_sound));
        authPreference = findPreference(getResources().getString(R.string.settings_auth_button));
    }

    @NonNull
    private boolean distanceListener(Preference preference, Object newValue) {
        Integer value = Math.min(LocationUtils.getEQUATOR(), Integer.parseInt((String) newValue));

        if (preference.equals(notificationDistPreference)) {
            preferences.setVisibleDistance(value);
        } else if (preference.equals(notificationAlarmPreference)) {
            preferences.setAlarmDistance(value);
        }
        update();
        return false;
    }

    @NonNull
    private boolean visibleListener(Preference preference, Object newValue) {
        preferences.putBoolean(preference.getKey(), (boolean) newValue);
        if (isAllHidden()) {
            ToastUtils.show(getActivity(), getString(R.string.no_one_accident_visible));
        }
        update();
        return false;
    }

    private boolean isAllHidden() {
        return !(preferences.getShowAcc() || preferences.getShowBreak() || preferences.getShowOther() || preferences.getShowSteal());
    }
}

File: Motocitizen/src/motocitizen/ui/fragments/SettingsFragment.kt
package motocitizen.ui.fragments

import android.os.Bundle
import androidx.preference.CheckBoxPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import motocitizen.content.AccidentsController
import motocitizen.datasources.preferences.Preferences
import motocitizen.main.R
import motocitizen.notifications.Messaging
import motocitizen.ui.Screens
import motocitizen.user.Auth
import motocitizen.user.User
import motocitizen.utils.EQUATOR
import motocitizen.utils.goTo
import motocitizen.utils.showToast

class SettingsFragment : PreferenceFragmentCompat() {
    private val PREFERENCES = R.xml.preferences

    // Инициализация будет происходить после onCreatePreferences
    private lateinit var buttonAuth: Preference
    private lateinit var buttonSound: Preference
    private lateinit var notificationSoundPreference: Preference
    private lateinit var authPreference: Preference

    private lateinit var notificationDistPreference: Preference
    private lateinit var notificationAlarmPreference: Preference
    private lateinit var showAcc: CheckBoxPreference
    private lateinit var showBreak: CheckBoxPreference
    private lateinit var showSteal: CheckBoxPreference
    private lateinit var showOther: CheckBoxPreference
    private lateinit var hoursAgo: Preference
    private lateinit var isTester: CheckBoxPreference
    private lateinit var maxNotifications: Preference
    private lateinit var useVibration: Preference

    private var login = Preferences.login

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(PREFERENCES, rootKey)

        // Привязка всех Preferences после инициализации
        buttonAuth = requirePreference(R.string.settings_auth_button)
        buttonSound = requirePreference(R.string.notification_sound)
        notificationSoundPreference = requirePreference(R.string.notification_sound)
        authPreference = requirePreference(R.string.settings_auth_button)

        notificationDistPreference = requirePreference("distanceShow")
        notificationAlarmPreference = requirePreference("distanceAlarm")
        showAcc = requireCheckBox("showAcc")
        showBreak = requireCheckBox("showBreak")
        showSteal = requireCheckBox("showSteal")
        showOther = requireCheckBox("showOther")
        hoursAgo = requirePreference("hoursAgo")
        isTester = requireCheckBox("isTester")
        maxNotifications = requirePreference("maxNotifications")
        useVibration = requirePreference("useVibration")

        setUpListeners()
        update()
    }

    private fun <T : Preference> requirePreference(key: String): T =
        findPreference(Preferences.getPreferenceName(key)) ?: error("Preference not found: $key")

    private fun <T : Preference> requirePreference(resId: Int): T =
        findPreference(getString(resId)) ?: error("Preference not found: ${getString(resId)}")

    private fun requireCheckBox(key: String): CheckBoxPreference =
        requirePreference(key)

    private fun Preference.onChangeListener(callback: (Preference, Any) -> Boolean) {
        onPreferenceChangeListener = Preference.OnPreferenceChangeListener(callback)
    }

    private fun Preference.onClickListener(callback: (Preference) -> Boolean) {
        onPreferenceClickListener = Preference.OnPreferenceClickListener(callback)
    }

    private fun update() {
        authPreference.summary = if (login.isNotEmpty()) "${User.roleName}: $login" else User.roleName
        maxNotifications.summary = Preferences.maxNotifications.toString()
        hoursAgo.summary = Preferences.hoursAgo.toString()
        notificationSoundPreference.summary = Preferences.soundTitle
        notificationDistPreference.summary = Preferences.visibleDistance.toString()
        notificationAlarmPreference.summary = Preferences.alarmDistance.toString()
        showAcc.isChecked = Preferences.showAccidents
        showBreak.isChecked = Preferences.showBreaks
        showSteal.isChecked = Preferences.showSteal
        showOther.isChecked = Preferences.showOther
        isTester.isChecked = Preferences.isTester
    }

    private fun setUpListeners() {
        buttonSound.onClickListener { soundButtonPressed() }
        buttonAuth.onClickListener { authButtonPressed() }

        notificationDistPreference.onChangeListener(::distanceListener)
        notificationAlarmPreference.onChangeListener(::distanceListener)
        maxNotifications.onChangeListener(::maxNotificationsListener)
        hoursAgo.onChangeListener(::hoursAgoListener)
        useVibration.onChangeListener { _, newValue -> vibrationListener(newValue) }
        isTester.onChangeListener(::isTesterListener)

        arrayOf(showAcc, showBreak, showOther, showSteal).forEach {
            it.onChangeListener(::visibleListener)
        }
    }

    private fun maxNotificationsListener(preference: Preference, newValue: Any): Boolean {
        preference.summary = newValue.toString()
        return true
    }

    private fun visibleListener(preference: Preference, newValue: Any): Boolean {
        when (preference.key) {
            "mc.show.acc" -> Preferences.showAccidents = newValue as Boolean
            "mc.show.break" -> Preferences.showBreaks = newValue as Boolean
            "mc.show.steal" -> Preferences.showSteal = newValue as Boolean
            "mc.show.other" -> Preferences.showOther = newValue as Boolean
        }
        if (isAllHidden) activity?.showToast(getString(R.string.no_one_accident_visible))
        update()
        return false
    }

    private fun isTesterListener(preference: Preference, newValue: Any): Boolean {
        Preferences.isTester = newValue as Boolean
        if (newValue) {
            Messaging.subscribeToTest()
        } else {
            Messaging.unSubscribeFromTest()
        }
        AccidentsController.resetLastUpdate()
        update()
        return false
    }

    private fun hoursAgoListener(preference: Preference, newValue: Any): Boolean {
        val value = if (newValue == "0") "1" else newValue.toString()
        preference.summary = value
        AccidentsController.resetLastUpdate()
        return true
    }

    private fun vibrationListener(newValue: Any): Boolean {
        Preferences.vibration = newValue as Boolean
        return true
    }

    private fun authButtonPressed(): Boolean {
        Auth.logout()
        requireActivity().goTo(Screens.AUTH)
        return true
    }

    private fun soundButtonPressed(): Boolean {
        parentFragmentManager.beginTransaction()
            .replace(android.R.id.content, SelectSoundFragment())
            .addToBackStack(null)
            .commit()
        return true
    }

    private fun distanceListener(preference: Preference, newValue: Any): Boolean {
        val value = (newValue as String).toInt().coerceAtMost(EQUATOR)
        when (preference) {
            notificationDistPreference -> Preferences.visibleDistance = value
            notificationAlarmPreference -> Preferences.alarmDistance = value
        }
        update()
        return false
    }

    private val isAllHidden: Boolean
        get() = !with(Preferences) { showAccidents || showSteal || showBreaks || showOther }
}
